### ISSUE TITLE
Java SDK seems to not play well with Replay tests. Setting child workflow ID using Temporal library seems to be throwing a non-deterministic error

### ISSUE THREAD

**Issue by gauravojha** [temporalio/sdk-java#2057]
Details attached on the temporal forum support request - https://community.temporal.io/t/workflowreplayer-replayworkflowexecution-throwing-nondeterministicexception-when-reading-workflow-id-from-workflow/12040/6

Copied from the post above

========

Here is a sample repo which replicates the error.
https://github.com/gauravojha/learningTemporal/tree/master

So what we are trying to do is

(Also mentioned here, hope this helps - [learningTemporal/src/main/java/dev/ojha/learningtemporal/workflows/HelloWorldWorkflowImpl.java at master · gauravojha/learningTemporal · GitHub](https://github.com/gauravojha/learningTemporal/blob/master/src/main/java/dev/ojha/learningtemporal/workflows/HelloWorldWorkflowImpl.java#L35-L40))

* Have a parent workflow trigger a child workflow
* What we are trying to do is, reusing the parent workflows ID (this we get by using the provided Temporal utility method - `Workflow.getInfo().getWorkflowId()`, and then using this is a prefix and just append "-child" to it. So for instance if the parent workflow ID is "helloWorld", we are assigning our child workflow ID as "helloWorld-child", as this simplifies some usecases of tracing and audit for our usecases
  * When we do this, and replay the payload through the replay tests, we get an undeterministic error, I have attached the trace for this specific case at the end
  * `Failure handling event 5 of type 'EVENT_TYPE_START_CHILD_WORKFLOW_EXECUTION_INITIATED' during replay. [TMPRL1100] Command COMMAND_TYPE_START_CHILD_WORKFLOW_EXECUTION doesn't match event EVENT_TYPE_START_CHILD_WORKFLOW_EXECUTION_INITIATED with EventId=5 on check workflowId with an expected value 'workflow_id_in_replay-child' and an actual value 'HelloWorldWorkflow-child'`
* However, if we don't reuse the parent's workflow ID, and let temporal use a random ID as the workflow ID of the child, then the replay tests pass.

**Comment by Quinn-With-Two-Ns** [comment]
How are you replaying history? If you are using `WorkflowReplayer.replayWorkflowExecution`  then when you [construct ](https://github.com/temporalio/sdk-java/blob/a41c64ece3876f584ebde70903668b817c5455d8/temporal-sdk/src/main/java/io/temporal/internal/common/WorkflowExecutionHistory.java#L60C10-L60C34)`WorkflowExecutionHistory` you need to specify the workflow ID

**Comment by gauravojha** [comment]
> How are you replaying history? If you are using `WorkflowReplayer.replayWorkflowExecution` then when you [construct ](https://github.com/temporalio/sdk-java/blob/a41c64ece3876f584ebde70903668b817c5455d8/temporal-sdk/src/main/java/io/temporal/internal/common/WorkflowExecutionHistory.java#L60C10-L60C34)`WorkflowExecutionHistory` you need to specify the workflow ID

@Quinn-With-Two-Ns sorry, i am not sure i understand.. I am not doing anything complicated in my code if you see it. my workflow implementation is pretty simple, it only triggers a child workflow. I am not doing any special replaying in there. Any replay is being done in the replay test which is written using instructions at https://docs.temporal.io/dev-guide/java/durable-execution#add-replay-test

This is the workflow implementation - https://github.com/gauravojha/learningTemporal/blob/652d2ba76e80b7d1a96946fae42b3d1dec0b3b3f/src/main/java/dev/ojha/learningtemporal/workflows/HelloWorldWorkflowImpl.java#L35-L40

This is the only place where any replay happens (in the WorkflowReplayer in a unit test) -  https://github.com/gauravojha/learningTemporal/blob/652d2ba76e80b7d1a96946fae42b3d1dec0b3b3f/src/test/java/dev/ojha/learningtemporal/TemporalTesting.java#L45

From the forum question where this is originally posted, it seems to be a bug in the SDK?

**Comment by gauravojha** [comment]
Got it.. Let me try @Quinn-With-Two-Ns, and thank you for pointing in the right direction .. i couldnt find this recommendation in the replay test documentation. If this is a critical statement, I feel this should be mentioned in the document in the document, since as a user I may not end up taking the first recommendation from the doc? 

> You need to specificy a workflow ID when using the replayer since it is not always in history

**Comment by Quinn-With-Two-Ns** [comment]
Yes I agree we should improve the documentation here. Passing the workflow ID is only important if the workflow uses the workflow ID as part of its logic which most don't.

**Comment by dmytrodanilenkov** [comment]
@Quinn-With-Two-Ns can you look to my PR?

**Comment by Quinn-With-Two-Ns** [comment]
@dmytrodanilenkov Yes I looked and approved CI to run yesterday and today.
